distributed/schedulerstate.py

```diff
@@ -0,0 +1,1419 @@
+        exception_blame = cython.declare(object, visibility='public')
+
+        suspicious = cython.declare(int, visibility='public')
+        retries = cython.declare(int, visibility='public')
+        nbytes = cython.declare(cython.Py_ssize_t, visibility='public')
+
+    def __init__(self, key, run_spec):
+        self.key = key
+        self.prefix = key_split(key)
+        self.run_spec = run_spec
+        self.state = None
+        self.exception = self.traceback = self.exception_blame = None
+        self.suspicious = self.retries = 0
+        self.nbytes = -1
+        self.priority = None
+        self.who_wants = set()
+        self.dependencies = set()
+        self.dependents = set()
+        self.waiting_on = set()
+        self.waiters = set()
+        self.who_has = set()
+        self.processing_on = None
+        self.has_lost_dependencies = False
+        self.host_restrictions = None
+        self.worker_restrictions = None
+        self.resource_restrictions = None
+        self.loose_restrictions = False
+
+    @cython.ccall
+    @cython.returns(cython.Py_ssize_t)
+    def get_nbytes(self):
+        nbytes = self.nbytes
+        return nbytes if nbytes >= 0 else DEFAULT_DATA_SIZE
+
+    @cython.ccall
+    @cython.locals(nbytes=cython.Py_ssize_t)
+    def set_nbytes(self, nbytes):
+        old_nbytes = self.nbytes
+        diff = nbytes - (old_nbytes or 0)
+        for ws in self.who_has:
+            ws.nbytes += diff
+        self.nbytes = nbytes
+
+    def __repr__(self):
+        return "<Task %r %s>" % (self.key, self.state)
+
+    def validate(self):
+        for cs in self.who_wants:
+            assert isinstance(cs, ClientState), (repr(cs), self.who_wants)
+        for ws in self.who_has:
+            assert isinstance(ws, WorkerState), (repr(ws), self.who_has)
+        for ts in self.dependencies:
+            assert isinstance(ts, TaskState), (repr(ts), self.dependencies)
+        for ts in self.dependents:
+            assert isinstance(ts, TaskState), (repr(ts), self.dependents)
+        # XXX
+        #validate_task_state(self)
+
+
+# -------------------------------------------------------------------------
+
+
+@cython.cclass
+class BaseSchedulerState(object):
+    """
+    """
+    if cython.compiled:
+        tasks = cython.declare(dict, visibility='public')
+        unrunnable = cython.declare(set, visibility='public')
+        task_duration = cython.declare(dict, visibility='public')
+
+        workers = cython.declare(object, visibility='public')
+        idle = cython.declare(object, visibility='public')
+        saturated = cython.declare(set, visibility='public')
+        total_occupancy = cython.declare(cython.double, visibility='public')
+
+    def __init__(self):
+        self.tasks = dict()
+        self.unrunnable = set()
+        self.task_duration = {prefix: 0.00001 for prefix in fast_tasks}
+        self.unknown_durations = defaultdict(set)
```

Comment:
You have to declare all attributes for a cclass. Suggests that there is probably something not working here.

Reply:
Actually, it's defined on `Scheduler` which derives `BaseSchedulerState`.

Reply:
If a base class uses it, why would a subclass define it?
Anyway, just be aware that Cython has to generate generic Python attribute access code in this case. You can easily see that with `cython -a`.

Reply:
Yeah, I understand that. This PR is an experiment, the goal is not to write well-architected code, but to judge how much Cython can improve performance on critical routines.